packages/ra-core/src/controller/EditController.tsx

```diff
@@ -91,123 +80,77 @@ interface EnhancedProps {
  *     );
  *     export default App;
  */
-export class UnconnectedEditController extends Component<
-    Props & EnhancedProps
-> {
-    componentDidMount() {
-        this.updateData();
-    }
+const EditController = (props: Props) => {
+    useCheckMinimumRequiredProps('Edit', ['basePath', 'resource'], props);
+    const translate = useTranslate();
+    const dispatch = useDispatch();
 
-    componentWillReceiveProps(nextProps: Props & EnhancedProps) {
-        if (
-            this.props.id !== nextProps.id ||
-            nextProps.version !== this.props.version
-        ) {
-            this.props.resetForm(REDUX_FORM_NAME);
-            this.updateData(nextProps.resource, nextProps.id);
-        }
-    }
+    const { basePath, children, id, resource, undoable } = props;
 
-    defaultRedirectRoute() {
-        return 'list';
-    }
+    const record = useSelector((state: ReduxState) =>
+        state.admin.resources[props.resource]
+            ? state.admin.resources[props.resource].data[props.id]
+            : null
+    );
 
-    updateData(resource = this.props.resource, id = this.props.id) {
-        this.props.crudGetOne(resource, id, this.props.basePath);
-    }
+    const isLoading = useSelector(
+        (state: ReduxState) => state.admin.loading > 0
+    );
 
-    save = (data, redirect) => {
-        const {
-            undoable = true,
-            startUndoable,
-            dispatchCrudUpdate,
-        } = this.props;
-        if (undoable) {
-            startUndoable(
-                crudUpdate(
-                    this.props.resource,
-                    this.props.id,
-                    data,
-                    this.props.record,
-                    this.props.basePath,
-                    redirect
-                )
-            );
-        } else {
-            dispatchCrudUpdate(
-                this.props.resource,
-                this.props.id,
-                data,
-                this.props.record,
-                this.props.basePath,
-                redirect
-            );
-        }
-    };
+    const version = useSelector(
+        (state: ReduxState) => state.admin.ui.viewVersion
+    );
 
-    render() {
-        const {
-            basePath,
-            children,
-            id,
-            isLoading,
-            record,
-            resource,
-            translate,
-            version,
-        } = this.props;
+    useEffect(() => {
+        dispatch(resetForm(REDUX_FORM_NAME));
```

Comment:
resetForm wasn't called on mount in the previous version. Isn't there a consequence of calling it here?

Reply:
I think we have e2e tests for this case but I will check